glue/viewers/image/state.py

```diff
@@ -572,6 +572,19 @@ def flip_limits(self):
         """
         self.attribute_lim_helper.flip_limits()
 
+    def set_slice(self, slices):
```

Comment:
I think this should not be a method to set the slices but instead we should have a callback property on image layer state which is whether to apply only to current slice or use whole data for limits, and if the former then it would use the slices from the viewer state. We should then listen to the viewer state slices and update the attribute limits helper.

Reply:
@astrofrog that property setting would then be passed as parameter to `attribute_lim_helper` (similar e.g. to `percentile`), and then create an updated `StateAttributeLimitsHelper` as needed? Or should the `ImageLayerState` property still use `self.attribute_lim_helper.set_slice` similar to the current method?

Reply:
I don't think it matters too much since ``StateAttributeLimitsHelper`` is private API, so I would go for what ends up being cleanest code-wise. However, since we don't always want ``StateAttributeLimitsHelper`` to listen to changes in the ``slice`` callback property (for example by default we don't), maybe keeping the ``self.attribute_lim_helper.set_slice`` approach would be easiest? (and fewest changes compared to what you have now)

Reply:
That's just the second question coming up now – just that switch would be added as callback (DDsomethingProperty), or `slices` as well? Because it seems once it is set to use current `slices`, it would have to act on any change of that, right?

Reply:
@dhomeier - we'd have to add a new callback property which would be the boolean switch. When that becomes ``True``, we could call ``add_callback`` which would watch for changes to ``slices``, and when the boolean switch becomes ``False`` it would call ``remove_callback``. Does that make sense?

Reply:
@astrofrog more or less what I figured out, too – just turned out that adding this to `ImageLayerState` did not seem to go anywhere, as `slices` is an `ImageViewerState` property. I could access it through `viewer_state`, but could not work out how to add that as a callback.
So setting the whole thing up in `ImageViewerState` as in 2e08899b works as it should, but now I have the problem of making sure I am really referencing the right level in `layers` (there's been only one in the examples so far).
Taking your point made in the Slack that this _should_ belong into `ImageLayerState`; I see if I added a new property there updated from `slices` (or directly on `numpy_slice_aggregation_transpose`, which is already used here), that could be set up as a callback – would that work?

Reply:
@astrofrog experimenting now with defining this via `@callback_property` decorator, but this, after some struggle to delay the initialisation from `numpy_slice_aggregation_transpose` now fails in glue-jupyter on
`TypeError: Object of type slice is not JSON serializable` 🤯
Should I go back the pedestrian way to construct the slice objects outside the callback_property, or is this whole approach doomed to fail?

Reply:
Pushing this mess for reference; this does not raise at least, but also does not update with changes to `self.viewer_state.slices`. 🤷‍♂️

Reply:
@dhomeier - regarding listening to changes in slices, have you tried doing the following from the layer state:

```
self.viewer_state.add_callback('slices', self._method_on_layer_state)
```

Does this not work?

Reply:
Whether in `__init__` or `_set_global_limits`, this gives a
```python
AttributeError: 'BqplotImageLayerState' object has no attribute '_method_on_layer_state'
```
(`self.viewer_state` does not have one either) – does this need to be defined yet, and what exactly should the function do?

Reply:
Just to be clear, ``_method_on_layer_state`` was a placeholder, I just mean you can use ``self.viewer_state.add_callback(...)``

Reply:
Excellent. Don't even need an extra `callback_property` then!